vyperlang/vyper
# Constant folding

### What I did
Add functionality for constant folding and replacing nodes within the AST.

### How I did it
* Add `evaluate` method to some Vyper node classes. This method returns a new node representing the literal value, or raises `InvalidType`.
* Add `Module.replace_in_tree` for replacing an existing node with a new one.
* Add `ast/folding.py` for high-level folding methods.

### How to verify it
Run the tests. I've added additional cases using `hypothesis` to verify that the folding outcome is the same as the EVM output.

### Cute Animal Picture
![image](https://user-images.githubusercontent.com/35276322/77969587-11403600-72fb-11ea-99c5-ec9fc1c0a02b.png)
